homeassistant/components/mqtt/sensor.py

```diff
@@ -271,7 +271,18 @@ def _update_state(msg: ReceiveMessage) -> None:
                 SensorDeviceClass.DATE,
                 SensorDeviceClass.TIMESTAMP,
             }:
-                self._attr_native_value = str(payload)
+                new_value = str(payload)
+                if self.device_class is not None and new_value == "":
```

Comment:
If you look at the `sensor` code, the condition for "must be numeric" is a little more complex:
- has device_class AND device_class is not in {custom, DATE, ENUM, TIMESTAMP}
- OR has unit of measurement
- OR has state class
- OR has precision

Maybe we need to add a property to the base sensor to indicate that a numeric value is expected.

Reply:
Let's keep this out of the beta scope for now.

Reply:
I agree we need that, but let's do that in a separate PR

Reply:
I think, if we want to introduce a breaking change, we should at least do it right. The other three cases should be taken into account.